File: medium/92. Reverse Linked List II/92. Reverse Linked List II.ts
// 92. Reverse Linked List II


class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
    }
}

const node5 = new ListNode(5)
const node4 = new ListNode(4, node5)
const node3 = new ListNode(3, node4)
const node2 = new ListNode(2, node3)
const node1 = new ListNode(1, node2)

function reverseBetween(head: ListNode | null, left: number, right: number): ListNode | null {
    if (!head || right === left) return head

    let first = new ListNode(0, head)
    // first.next = head
    let prev = first // position after which need to flip the list

    for (let i = 0; i < left - 1; i++) {
        prev = prev.next!
    }

    let curr = prev.next // left position in the list

    for(let i = 0; i < right - left; i++) {
        if(!curr || !curr.next) break // break if curr or curr.next is empty

        let next = curr.next // choose next item. left + 1
        curr.next = next.next // left + 1 = left + 1 + 1
        next.next = prev.next // left + 1 + 1 = left
        prev.next = next
    }

    return first.next

};

console.log(reverseBetween(node1, 2, 4));


export { }

File: medium/92. Reverse Linked List II/92. Reverse Linked List II.py
# 92. Reverse Linked List II
# Дано: односвязный список элементов, а также значения left и right, между которыми нужно перевернуть порядок элементов
# Условие: вернуть связанный список с перевернутым порядком элементов в некоторых позициях.

# Пример 1:
# Вход: head = [1,2,3,4,5], left = 2, right = 4
# Выход: [1,4,3,2,5]

from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


node_7 = ListNode(7, None)
node_6 = ListNode(6, node_7)
node_5 = ListNode(5, node_6)
node_4 = ListNode(4, node_5)
node_3 = ListNode(3, node_4)
node_2 = ListNode(2, node_3)
node_1 = ListNode(1, node_2)


class Solution:
    def reverseBetween(
        self, head: Optional[ListNode], left: int, right: int
    ) -> Optional[ListNode]:
        # last_elements: ListNode | None = head
        # last_index = 1
        # while last_elements:
        #     if last_index <= right:
        #         last_elements = last_elements.next
        #         last_index += 1
        #     else:
        #         break

        # first: ListNode | None = ListNode(head.val)  # First element of the lined list
        # prev: ListNode | None = (
        #     None  # Reversed list with elements that have starting on the left and ending on the right
        # )
        # curr: ListNode = head  # Just current element during while cycle
        # index = 1  # For controlling position between left and right

        # while curr:
        #     next: ListNode | None = curr.next
        #     if index >= left and index <= right:
        #         curr.next = prev or last_elements
        #         prev = curr
        #         curr = next
        #         index += 1
        #     else:
        #         if left == 1:
        #             first = prev
        #         else:
        #             first.next = prev
        #         break

        #     # if not first:
        #     #     curr.next = None
        #     #     first = curr
        #     #     curr = next
        #     #     index += 1
        #     # else:
        #     #     if index <= right and index >= left:
        #     #         curr.next = prev or last_elements
        #     #         prev = curr
        #     #         curr = next
        #     #         index += 1
        #     #     elif index > right:
        #     #         if not first.next:
        #     #             first.next = prev
        #     #         break

        if not head or left == right:
            return head

        first = ListNode(0)
        first.next = head
        prev = first

        for _ in range(left - 1):
            prev = prev.next

        curr = prev.next
        next_val = None

        for _ in range(right - left):
            if not curr or not curr.next:
                break

            next_val = curr.next  # link on the next item
            curr.next = next_val.next or None
            next_val.next = prev.next
            prev.next = next_val

        return first.next


solution = Solution()

items = []
res = solution.reverseBetween(node_1, 2, 4)
while res:
    items.append(res.val)
    res = res.next
print(items)
